src/neo/Cryptography/MPT/MPTTrie.Proof.cs

```diff
@@ -0,0 +1,79 @@
+using Neo.IO;
+using Neo.Persistence;
+using System;
+using System.Collections.Generic;
+
+namespace Neo.Cryptography.MPT
+{
+    public partial class MPTTrie<TKey, TValue>
+        where TKey : notnull, ISerializable, new()
+        where TValue : class, ISerializable, new()
+    {
+        public bool GetProof(TKey key, out HashSet<byte[]> set)
+        {
+            set = new HashSet<byte[]>(ByteArrayEqualityComparer.Default);
+            var path = ToNibbles(key.ToArray());
+            if (path.Length < 1) return false;
+            return GetProof(ref root, path, set);
+        }
+
+        private bool GetProof(ref MPTNode node, ReadOnlySpan<byte> path, HashSet<byte[]> set)
+        {
+            switch (node)
+            {
+                case LeafNode leafNode:
+                    {
+                        if (path.Length < 1)
+                        {
+                            set.Add(leafNode.Encode());
+                            return true;
+                        }
+                        break;
+                    }
+                case HashNode hashNode:
+                    {
+                        if (hashNode.IsEmpty) break;
+                        var newNode = Resolve(hashNode);
+                        if (newNode is null) break;
+                        node = newNode;
+                        return GetProof(ref node, path, set);
+                    }
+                case BranchNode branchNode:
+                    {
+                        set.Add(branchNode.Encode());
+                        if (path.Length < 1)
+                        {
+                            return GetProof(ref branchNode.Children[BranchNode.ChildCount - 1], path, set);
+                        }
+                        return GetProof(ref branchNode.Children[path[0]], path[1..], set);
+                    }
+                case ExtensionNode extensionNode:
+                    {
+                        if (path.StartsWith(extensionNode.Key))
+                        {
+                            set.Add(extensionNode.Encode());
+                            return GetProof(ref extensionNode.Next, path[extensionNode.Key.Length..], set);
+                        }
+                        break;
+                    }
+            }
+            return false;
+        }
+
+        public static TValue VerifyProof(UInt256 root, TKey key, HashSet<byte[]> proof)
+        {
+            var memoryStore = new MemoryStore();
+            foreach (byte[] data in proof)
+            {
+                memoryStore.Put(MPTDb.Prefix, Crypto.Hash256(data), data);
+            }
+            ISnapshot snapshot = memoryStore.GetSnapshot();
+            var trie = new MPTTrie<TKey, TValue>(root, snapshot);
+            foreach (byte[] data in proof)
+            {
+                snapshot.Put(MPTDb.Prefix, Crypto.Hash256(data), data);
+            }
```

Comment:
Why put twice?

Reply:
Everything is fine now, except this one.

Reply:
😅 My mistake.